### DISCUSSION TITLE
Integrating React Native

### DISCUSSION THREAD

**Discussion by LuigiZaccagnini** [Seneca-CDOT/telescope#2667] [Ideas]
Hello everyone! After the meeting today I wanted to start a discussion on how we can move forward with the development of the React Native app! Things we should discuss are:

- What tools/CLI should we use for React Native?
- What should our goal be for 3.0 release?
- How should we break down this feature into issues?
- Everyone's experience with React Native? 
- How can we get started?

I have discussed some these in my [blog](https://dev.to/luigizaccagnini/getting-telescopes-react-native-app-2h1a) but wanted to know what the community thought. I also need help starting the project with the community! Once we have things rolling we can create a project in the GitHub for future organization.

**Comment by chrispinkney** [comment]
Should we also consider moving the native port to another repo? ex: `Seneca-CDOT/telescope-mobile`? I think the main repo will get really messy if everything will be contained in one, not to mention build tools, configs, etc.

**Reply by LuigiZaccagnini** [reply]
Yeah I agree! I think there should be another repository because it might make this repository crazy to navigate.

**Reply by humphd** [reply]
Monorepo all the way.

The reason you want to do this in the same repo is that you'll be able to update back-end APIs, tooling, and front-end in a single PR.  Google has *all* its code in one repo for this reason.

Navigating the boundaries of tons of smaller repos is harder than moving between directories in a monorepo.

Stick this in `src/mobile` and reap the rewards.

**Reply by LuigiZaccagnini** [reply]
Never knew that Google had all of its code in one repo! You make a great point, so now we know where to start!

**Comment by nguyenhung15913** [comment]
I would highly suggest using React Native CLI rather than Expo CLI. The reason is when I was working with an Uber Clone Project, I used some libraries related to the map such as "react-native-geolocation-service", "react-native-maps", etc. And somehow some of their installation's documentations only pointed towards React Native CLI but not Expo CLI. This made me have to delete my Expo CLI project and moved to React Native CLI. But honestly, in my opinion, Expo CLI is way more easier to use than  React Native. This is just my experience :D and so far I have just touch React Native for one time.

**Reply by humphd** [reply]
I think it happens when you want/need to use native components.

**Reply by nguyenhung15913** [reply]
That is true. One of the issues was related to Linking and Creating the permission (Allowing location or camera).

**Reply by LuigiZaccagnini** [reply]
Yeah Expo does not support native modules but I was wondering if we needed native modules? 
These are the pros and cons I found:

## Expo CLI

### Pros:
- Fast development setup
- Quickest way to start development with React Native
- Live reload in Development
- Library linking
- Can test app on apple/android devices (Expo builds apps)
- Can eject/convert back into a React Native CLI

### Cons:
- Builds are done by Expo
- Native modules are not supported
- Expo apps are bigger because all libraries are included

## React Native CLI

### Pros:
- Builds are done locally
- Native modules are supported
- You have more control over app size

### Cons:
- Slow development setup
- No live reload in Development
- No library linking (things need to be linked with react-native link)
- Can't turn into an expo app

Maybe these can help choose what would best fit the project?

**Reply by HyperTHD** [reply]
Would probably be best to consider what modules we'd need, and whether they're native or not. A decent list would be a good start in determining which cli to go with.

**Reply by humphd** [reply]
If Expo doesn't negate us moving to native later, I'm not sure it matters.  We could start with Expo and migrate when it's needed.

I'd pick whatever gets the majority of us up and running the fastest.  It sounds like that's Expo?  But I haven't done any RN so help me out.

**Reply by LuigiZaccagnini** [reply]
Yes Expo allows us to eject the project back to native!

**Comment by Kevan-Y** [comment]
For the UI, maybe we can of use some UI library to make it a bit easier for css, such as [React Native Elements](https://reactnativeelements.com/), or [Native Base](https://nativebase.io/).

**Reply by LuigiZaccagnini** [reply]
Native Base look really good! I have used React Native Elements and it is also not a bad choice.
